README.md
# Mopslig - My Open Source License Generator 

This project can be used to generate license keys, verifycation hashes,
extraction "pseudo" keys, extraction hashes and a license object,
in which all license types are stored.
Basically you will get a list of generated 'license keys' and some binary files
which are able to tell you if your license key is valid, generate some licence for 
your clients and other stuff. Demo will be also available soon.

## Prerequisites
- pp - perl PAR packager
- perl Path::Tiny
- perl JSON::XS
- perl Getopt::Long
- perl Time::Piece
- perl Crypt::CBC
- perl Digest::SHA
- perl Digest::SHA1
- perl Crypt::PBKDF2
- perl Crypt::PBKDF2::Hash::HMACSHA2
- perl Crypt::Rijndael


## Usage - kinda automatic
- edit config.json
- run build_all.sh
```sh
$ cd bin
$ ./build_all.sh -y
```
- pick one of generated "serial numbers" from data/serials.txt 
- mopslig generates allways one default START-DE12-FA34-UL56-T789 key, because i need it (don't ask why)
- save one of those generated keys to key.lic file
- run ./test_it.pl
```sh
$ cd bin
$ ./test_it.pl
```
See [binaries docmentation usage](https://github.com/jdrab/mopslig/blob/master/mopslig.md) for more informations.

## How do i remove this mess?
- all generated files including key.lic can be removed by ./remove_files.sh, just run 
```sh
$ cd bin
$ ./remove_files.sh -y
```
or run these commands from the `bin` directory 

```sh
  rm -rf ../data/*
  unlink ../tmp/imprint.lic
  unlink ../tmp/client.lic
  unlink ../tmp/key.lic
  rm -rf ../tmp/*.tmp
  unlink ../build/client-lic
  unlink ../build/validate-lic
  unlink ../build/verify-key
  unlink ../build/imprint-lic
```

## If you don't like automagic
  - run  mopslig.pl; it will generate licensing keys, hashes, extraction keys and extraction hashes.
  - key - is used as main license key for client
  - key_hash - is generated for key and is used to verify that key is valid. (PBKDF2 HMACSHA2 512)
  - extraction_key - is extracted from key by a retarded function (but hey, it works)
  - extraction_hash - this is the same as `key_hash` is to the `key`, except it is generated for extraction_key


```sh
$ cd bin
$ ./mopslig.pl sure
```

- build binary files in the same `bin` directory
```sh
$ ./build-client-lic.pl
$ ./build-imprint-lic.pl
$ ./build-validate-lic.pl
$ ./build-verify-key.pl 
 ```
- pick one of generated serial numbers from ../data/serials.txt 
- generate client.lic and check exit code. You can generate client license file to other file too,
but you will have to change test_it.pl later. 

```sh 
$ cd build
$ ./client-lic --generate --key YOUR_LICENSE_KEY --output client.lic 
```

- validate license file ./validate-lic, check exit code 

```sh 
$ # run from `build` directory
$ ./validate-lic --validate
```

- read imprint.lic file

```sh
$ # run from `build` directory
$ ./imprint-lic --key YOUR_LICENSE_KEY --imprint-file path_to_imprint.lic
```

### Files

#### key.lic 
> your license key (mopslig.pl generates one default key - START-DE12-FA34-UL56-T789)
You don't need to save your license key in this file but binary `validate-lic` requires path to this file. 
It reads key.lic content and uses it as license key in validation
process.

#### config.json
>JSON file with informations about package names and key
```json
{
  "licenses": {
    "valid_since": "2016-08",
      "valid_until": "2017-07",
    "amounts": {
      "start": 10,
      "plus":  10,
      "premium": 10
    },
    "types": {
      "start": {
        "modulename": {
          "license_amount": 5,
          "valid_since": "2016-08",
          "valid_until": "2017-07"
        },
        "valid_since": "2016-08",
        "valid_until": "2017-07"
      },
      "plus": {
        "modulename": {
          "license_amount": 20,
          "valid_since": "2016-08",
          "valid_until": "2017-07"
        },
        "valid_since": "2016-08",
        "valid_until": "2017-07"
      },
      "premium": {
        "modulename": {
          "license_amount": 50,
          "valid_since": "2016-08",
          "valid_until": "2017-07"
        },
        "valid_since": "2016-08",
        "valid_until": "2017-07"
      }            
    }
  }
}
```

#### mopslig.pl
> will generate license keys,verification hashes, extraction keys 
and extraction hashes for extracting license files from licenses object

```sh
$ ./mopslig.pl 

Are you sure?

Your files will be OVERWRITEN.
Settings are read from ./config.json file.
If you're really sure you want to generate NEW serial numbers and hashes, please say 'sure'.

Usage:   ./mopslig.pl sure      --no-dots       --debug

```

#### build-client-lic.pl 
> Reads the build-id file, config.json. Strings MOPSLIG_BUILD_ID and MOPSLIG_LICSENSE_CONFIG are repalced
by values from those two files.

#### client-lic.template
> Almost complete perl file. Only sompe parts are replaced with files generated in the `data` folder.

#### client-lic.tmp
> This file is generated by `./build-client-lic.pl` - this is the full perl script source.
It is used by the perl pp command to build `client-lic` binary.

#### client-lic
> Binary file built from client-lic.tmp by `pp` command

#### build-imprint-lic.pl 
> This script will replace the string MOPSLIG_BUILD_ID in file imprint-lic.template with content of `data/build-id` file.

#### imprint-lic.template
> Almost complete perl file. Only sompe parts are replaced with files generated in the `data` folder.

#### imprint-lic.tmp
> This file is generated by `./build-imprint-lic.pl` - this is the full perl script source.
It is used by the perl pp command to build `imprint-lic` binary.

#### build-validate-lic.pl 
> This script will replace the strings MOPSLIG_BUILD_ID and MOPSLIG_EXTRACTION_HASHES in file validate-lic.template 
with content of `data/build-id` and `data/extraction-hashes.txt` files.

#### validate-lic.template
> Almost complete perl file. Only sompe parts are replaced with files generated in the `data` folder.

#### validate-lic.tmp
> This file is generated by `./build-validate-lic.pl` - this is the full perl script source.
It is used by the perl pp command to build `validate-lic` binary.

#### build-verify-key.pl 
> This script will replace the strings MOPSLIG_BUILD_ID and MOPSLIG_VERIFY_HASHES in file verify-key.template 
with content of `data/build-id` and `data/verify-hashes.txt` files.

#### verify-key.template
> Almost complete perl file. Only sompe parts are replaced with files generated in the `data` folder.

#### verify-key.tmp
> This file is generated by `./build-verify-key.pl` - this is the full perl script source.
It is used by the perl pp command to build `verify-key` binary.


### Versions

- Version 0.2.0 - 2016-08-26
- Version 0.1.0 - 2016-08-23
- Version 0.0.2 - 2016-08-03
- Version 0.0.1 - 2016-08-01

see [CHANGELOG](https://github.com/jdrab/mopslig/blob/master/CHANGELOG.md) for more info

### Todos
  
  - Testst,tests and moar tests..
  - Improve documentation
  - ~~Add support for --imprint-to in validate-lic~~
  - Add support for --no-default-key to mopslig.pl 
  - live demo using lxd
  - web interface (for demo & lazy me)




CHANGELOG.md
## [0.2.0] - 2016-08-26
### Changed
- directory lib/ moved to bin/lib - FIX issue #1
- usage output of binaries reformated
- README.md update (perl dependencies)

### Added
- docs for generated binary files - mopslig.md

## [0.1.0] - 2016-08-23
### Changed
- complete change of directory structure
- Added Path::Tiny dependency :)

### Added
- Changelog created :D
- support for --imprint-output in validate-lic

### Removed
- Removed File::Slurp dependency

## [0.0.2] - 2016-08-03

## [0.0.1] - 2016-08-01

validate-lic.pl
#!/usr/bin/perl env
use warnings;
use strict;
use Getopt::Long;
use File::Slurp;
use Digest::SHA1;

use Crypt::PBKDF2;
use Crypt::PBKDF2::Hash::HMACSHA2;
use Crypt::CBC;
use Crypt::Rijndael;

use File::Basename qw(dirname);
use Cwd qw(abs_path);
use lib dirname( abs_path $0) . '/lib';

use Mopslig::Generator;
use Data::Dumper;

use JSON::XS;

my $key_file = './key.lic';
my $lic_file = './client.lic';

GetOptions(
    'serial'      => \my $serial,
    'for-module'  => \my $module,
    'get-amount'  => \my $get_amount,
    'valid_until' => \my $valid_until,
    'valid-since' => \my $valid_since,
    'amount'      => \my $amount,

    #'output=s'=> \my $output,
    'build-id' => \my $build_id
);

my $build_id_val = "MOPSLIG_BUILD_ID";

if ($build_id) {
    print $build_id_val. "\n";
    exit;
}
######
#
#	EXIT CODES
# 2 	key file does not exist
# 3 	license file doest not exits
#
######

unless ($serial) {
    unless ( -e $key_file ) {
        exit(2);
    }
    $serial = File::Slurp::read_file($key_file);
}

chomp($serial);

#FIXME: print usage

my @hashes = qw(
    MOPSLIG_EXTRACTION_HASHES
);

my $hash_class = 'HMACSHA2';
my $sha_size   = 512;
my $salt_len   = 10;
my @balik_data;
my $prefix = '{X-PBKDF2}HMACSHA2+512:AAAD6A:';

my $pbkdf2 = Crypt::PBKDF2->new(
    hash_class => $hash_class,
    hash_args  => { sha_size => $sha_size },
    salt_len   => $salt_len
);

sub validate {
    my $key = shift;
    my $i   = 0;
    my $sum = scalar(@hashes);

    foreach my $b1 (@hashes) {
    	print ".";
        if ( $pbkdf2->validate( $prefix . $b1, $key ) ) {
            return $key;
        }
    }
    return 0;
}
print "Serial: $serial\n";
# zo $serialu potrebujem extraction kluc vygenerovat
my $extraction = Mopslig::Generator::generate_key_for_lic_extraction($serial);
if ( validate($extraction) ) {
    my $cipher = Crypt::CBC->new(
        -key    => $extraction,
        -cipher => "Crypt::Rijndael"
    );
    unless( -f $lic_file) { 
    	exit(3);
    }

	my $enc_lic = File::Slurp::read_file($lic_file, binmode => ':raw');

    my $dec = $cipher->decrypt($enc_lic);

    print $dec;
# {
#           'valid_since' => '2016-08',
#           'kbsusers' => {
#                           'valid_since' => '2016-08',
#                           'valid_until' => '2017-07',
#                           'license_amount' => 5
#                         },
#           'valid_until' => '2017-07',
#           'key' => 'START-DE12-FA34-UL56-T789'
#         };

}
else {
    print "INVALID\n";
    exit(1);
}

exit;

# my $products     = qq(MOPSLIG_LICSENSE_CONFIG);
# my $product_data = JSON::XS->new->utf8->decode($products);

# my $key_lic = -e $key_file ? File::Slurp::read_file($key_file);

# my $lic
#     = $product_data->{licenses}{types}{ lc( ( split( '-', $key_lic ) )[0] ) };

# if ($refresh) {
#     my ( undef, undef, undef, undef, $month, $year ) = localtime;
#     $month += 1; # lebo mesiace oznacuje od 0 do 11
#     $year += 1900;

#     my $valid_until = $lic->{valid_until};

#     my ( $valid_year, $valid_month ) = split( "-", $valid_until );

#     if ( $valid_year >= $year && $valid_month >= $month ) {
#         exit(0);
#     }
# }

# $lic->{key} = $key_lic;

# my $ekey = Mopslig::Generator::generate_key_for_lic_extraction($key_lic);

# my $lic_json = JSON::XS->new->encode($lic);

# use Crypt::CBC;
# use Crypt::Rijndael;

# my $cipher = Crypt::CBC->new(
#     -key    => $ekey,
#     -cipher => "Crypt::Rijndael"
# );

# my $enc = $cipher->encrypt($lic_json);
